Dockerize

jhsu802701 opened the pull request:
# 💅🏼 What issue does this fix?
Issue #576 .  The docker/build script sets up the Dockerfile (filling in the Ruby version), sets up the Docker containers, runs all tests (Rails and Yarn), seeds the database, runs RuboCop, and runs Brakeman.  The docker/git_check script is handy for running all necessary tests and checks before making Git commits and pushes.

katelovescode reviewed (changes requested):
Getting a couple errors/bad outputs while running docker/build:

```
Configure .docker-env/development/database
sed: 1: ".docker-env/development ...": invalid command code .
sed: 1: ".docker-env/development ...": invalid command code .
Configure init.sql
sed: 1: "init.sql": command i expects \ followed by text
sed: 1: "init.sql": command i expects \ followed by text
sed: 1: "Dockerfile": extra characters at the end of D command
```

```
Building rails
Step 1/9 : FROM ruby:RUBY_VERSION
Service 'rails' failed to build : manifest for ruby:RUBY_VERSION not found: manifest unknown: manifest unknown
--------------------------
BEGIN: yarn install (root)
--------------------------
Building rails
Step 1/9 : FROM ruby:RUBY_VERSION
Service 'rails' failed to build : manifest for ruby:RUBY_VERSION not found: manifest unknown: manifest unknown
-------------------
docker-compose down
Removing network pieforproviders_default
----------------------------------------------------------------------------------------------
mkdir -p /home/winner/myapp/tmp/bundle && chown -R winner:winner /home/winner/myapp/tmp/bundle
Creating network "pieforproviders_default" with the default driver
Building rails
Step 1/9 : FROM ruby:RUBY_VERSION
Service 'rails' failed to build : manifest for ruby:RUBY_VERSION not found: manifest unknown: manifest unknown
--------------------------
BEGIN: yarn install (root)
--------------------------
Building rails
Step 1/9 : FROM ruby:RUBY_VERSION
```

jhsu802701 commented on README.md:
```diff
@@ -26,6 +26,41 @@ Learn more at [www.pieforproviders.com](http://www.pieforproviders.com)
+### Other Important Scripts
+* Enter the command "docker/git_check" before "git add" and "git commit".  This runs the tests, Rubocop, and Brakeman.  The docker/git_check script is a sanity check to allow you to make sure to commit quality working code only.
+* Enter "docker/nuke" to destroy the Docker image, container, and networks.
+* Enter "docker/nuke" to destroy the Docker container but leave the base images in place.
```
Kate, it looks like the "sed" command does not work on your machine.  Are you using a Mac?  (I'm primarily a Linux user.)  If so, I can try out my branch on my Mac and come up with a solution that allows the docker/build script to work in Mac OS environments.

jhsu802701 replied:
I corrected the typo regarding the "docker/nukec" command.

katelovescode replied:
Yes, I'm on a Mac! Let me know if you need me to investigate anything.

jhsu802701 commented on docker-compose.yml:
```diff
@@ -0,0 +1,64 @@
+    container_name: pie_rails
+    build: .
+    volumes:
+      - .:/home/winner/myapp
```
* /home/winner/myapp is the WORKDIR specified in the Dockerfile (Dockerfile-template).
* "winner" is the username in the Docker environment.  I prefer to keep the username and the project name different, because using the same thing for both can lead to confusion if I need to troubleshoot/debug something.
* I set up the Rails Docker container to be run by the regular user as default.  Doing development as root is generally frowned on, yet it seems that most uses of Docker involve running the Docker containers as root.
* I keep the working directory in the Docker container one level below the user home directory, because making the working directory and user home directory one and the same contaminates the repository with user configuration files and directories.

katelovescode replied:
out of curiosity is `/home/username` the typical setup for Linux boxes?  For some reason I thought it was `/usr/username` or `/username` but I might be misremembering.  I understand wanting to keep it separate for debugging, can we use something like `docker` so it's clearer where it's coming from rather than `winner`?  I'm fine if you want to keep the run-docker-as-regular-user behavior though, that's fine.

jhsu802701 replied:
Yes, /home/username is the typical Linux setup.  I don't want to use "docker" as the username, because that could be a source of confusion.  If you wish, I could change the username from "winner" to something like "docker1" or "docker_user".

jhsu802701 commented on docker-compose.yml:
```diff
@@ -0,0 +1,64 @@
+    volumes:
+      - .:/home/winner/myapp
+    environment:
+      - BUNDLE_PATH=/home/winner/myapp/tmp/bundle
```
Using the tmp/bundle directory as BUNDLE_PATH means that the "bundle install" process is much quicker after you run the "docker/nuke" or "docker/nukec" command because the "bundle install" products are saved in the host environment.  This speeds up the debugging/troubleshooting process, which has been essential for working on dockerization.

katelovescode replied:
Ya! this was more about changing the `winner` name, so if you go with the username suggestions above, that's what this was about.  Sorry for the confusing comment :)

jhsu802701 commented:
So far, I've addressed all the issues except for the "sed" issue.  I'll try out this branch on my Mac computer.  Assuming that this is a Mac issue, I'll probably go with one of these options:
* At the beginning of the docker/build script, add a directive to Mac users to install gsed followed by "Press Enter to continue."
* In the docker/build script, add an if statement that automatically installs gsed on MacOS systems if it's not already in place.
* In the docker/replace_sif script, use an if/then/else command to create an alternate MacOS path.

jhsu802701 commented:
I have made all of the requested changes to this branch.  The sed commands now work in MacOS as well as in Linux.  I have removed all changes to the db/schema.rb and pie_erd.pdf files.

Are any other changes needed for a merge?

katelovescode commented:
> Also, I'll put the original db/schema.rb and pie_erd.pdf files back into the branch after I've made all of the other changes required. (For some reason, running the migrations changes these files.)

The reason, I think, is just because there aren't currently any data migrations in that directory so your local install of the `data-migrate` gem thinks there's no reason to generate the data migrations table yet (🙄 ) - the first time you have a data migration, that table will be created and will stay.  It might make sense for us to add it retroactively to the first schema migration so it doesn't get lost.  I'll take a look at that.

katelovescode commented:
> I have made all of the required changes in this branch. The sed commands now work in MacOS as well as in Linux. I have removed all changes to the db/schema.rb and pie_erd.pdf files.
> 
> Are any other changes needed for a merge?

Gonna give this a run tonight and I'll keep you posted!

katelovescode reviewed (changes requested):
Commented on the name changes throughout, also would love a lightweight script to run the server without as many prep steps; often when doing dev I have to do that manually anyway as a part of the process and waiting 5 mins for the server to come up is a development/context switching burden.

jhsu802701 commented:
I changed the username from "winner" to "docker_user", changed "myapp" to "pie, and added docker/qserver for the quick version of docker/server that skips the "bundle install and database migration steps.

katelovescode reviewed (approved):
Looks great, thanks!
